driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,401 @@
+        childPolicy.onDown(host);
+    }
+
+    @Override
+    public void onRemove(Host host) {
+        childPolicy.onRemove(host);
+    }
+
+    /**
+     * Creates a new error aware policy builder given the child policy
+     * that the resulting policy should wrap.
+     *
+     * @param childPolicy the load balancing policy to wrap with error
+     *                    awareness.
+     * @return the created builder.
+     */
+    public static Builder builder(LoadBalancingPolicy childPolicy) {
+        return new Builder(childPolicy);
+    }
+
+    /**
+     * Closes the policy. Calls the child policy's {@link LoadBalancingPolicy#close()} method, and shuts down the updater
+     * service created for the update of the {@link ErrorAwarePolicyTracker}.
+     */
+    @Override
+    public void close() {
+        childPolicy.close();
+        updaterService.shutdown();
+    }
+
+    static class Builder {
+        private final LoadBalancingPolicy childPolicy;
+
+        private ErrorAwarePolicyTracker errorTracker;
+        private long updatePeriod = 1000;
+        private int errorsPerSecThreshold = 10;
+        private int retryPeriod = 5000;
+
+        public Builder(LoadBalancingPolicy childPolicy) {
+            this.childPolicy = childPolicy;
+        }
+
+        public Builder withUpdatePeriodMillis(long updatePeriodMillis) {
+            this.updatePeriod = updatePeriodMillis;
+            return this;
+        }
+
+        // Hosts that have a number of errors/sec higher than the threshold are ignored.
+        public Builder withInclusionThreshold(int errorsPerSecThreshold) {
+            this.errorsPerSecThreshold = errorsPerSecThreshold;
+            return this;
+        }
+
+        public Builder withErrorTracker(ErrorAwarePolicyTracker errorTracker) {
+            this.errorTracker = errorTracker;
+            return this;
+        }
+
+        public Builder withRetryPeriod(int retryPeriod) {
+            this.retryPeriod = retryPeriod;
+            return this;
+        }
+
+        public ErrorAwarePolicy build() {
+            return new ErrorAwarePolicy(childPolicy, errorTracker, updatePeriod, errorsPerSecThreshold, retryPeriod);
+        }
+    }
+
+    public static class PerHostErrorTracker implements ErrorAwarePolicyTracker {
+
+        private Map<Host, Integer> errors = new HashMap<Host, Integer>();
+
+        private Map<Host, ErrorRecord> errorsPerSec = new HashMap<Host, ErrorRecord>();
+
+        private long retryPeriod;
+
+        private int errorsPerSecThreshold;
+
+        private double errorsThresholdRate;
+
+        private long updatePeriod;
+
+        @Override
+        public void notifyError(Host host, Statement statement, Exception exception) {
+            // should just keep counts of errors for each hosts
+            if (!shouldAccountForError(host, statement, exception)) {
+                return;
+            }
+            Integer entry;
+            Integer count = errors.get(host);
```

Comment:
Should use a concurrent map implementation here to avoid possible undercounting.